[libc] Support _IONBF buffering for read_unlocked

jackhong12 opened the pull request:
Support _IONBF buffering for read_unlocked. Add the functions read_unlocked_nbf() and read_unlocked_fbf().

Fixes: #120155

jackhong12 committed:
[libc] Support _IONBF buffering for read_unlocked

- Add the functions read_unlocked_nbf and read_unlocked_fbf.

jackhong12 committed:
[libc] Copy the data from the buffer first.

For read_unlocked() with the _IONBUF mode, there is still an
one-character buffer which is used to store the data inserted by
ungetc(), so we always need to check the buffer first.

michaelrj-google commented on libc/src/__support/File/file.cpp:
```diff
@@ -190,6 +190,17 @@ FileIOResult File::read_unlocked(void *data, size_t len) {
+  } else if (bufmode == _IOFBF) { // fully buffered
+    return read_unlocked_fbf(static_cast<uint8_t *>(data), len);
+  } else /*if (bufmode == _IOLBF) */ { // line buffered
+    // There is no line buffered mode for read. Use fully buffer instead.
```
spelling nit:
```suggestion
    // There is no line buffered mode for read. Use fully buffered instead.
```

michaelrj-google commented on libc/src/__support/File/file.cpp:
```diff
@@ -190,6 +190,17 @@ FileIOResult File::read_unlocked(void *data, size_t len) {
+  }
+}
+
+FileIOResult File::copy_data_from_buf(uint8_t *data, size_t len) {
```
since this can't ever return an error, this should return `size_t` instead of `FileIOResult`.

michaelrj-google commented on libc/src/__support/File/file.cpp:
```diff
@@ -209,12 +220,22 @@ FileIOResult File::read_unlocked(void *data, size_t len) {
+
   if (to_fetch > bufsize) {
     auto result = platform_read(this, dataref.data(), to_fetch);
     size_t fetched_size = result.value;
```
I can't put a comment on it because it's not a bug in your code, it's a bug in the code I wrote, but it needs to be fixed nonetheless: On line 226 of the original file it has `return {available_data + fetched_size, result.has_error()};` which is incorrect, it should be `return {available_data + fetched_size, result.error};`. If you could fix that in this patch that would be much appreciated.

jackhong12 replied:
Got it. I also change the line 286 from `result.has_error()` to `result.error`. I think it has the same problem.

michaelrj-google replied:
sounds good, I must've missed that one.

michaelrj-google reviewed (commented):
Overall LGTM, just a couple things to clean up before landing. Thanks for picking this up!

jackhong12 committed:
[libc] Change the code according to the review

- The function copy_data_from_buf returns size_t.
- Fix the spelling error.
- Fix the incorrect data, error, in FileIOResult.

jackhong12 committed:
[libc] Use the specific data type FileIOResult

jackhong12 committed:
[libc] Format the change

michaelrj-google reviewed (approved):
LGTM, thanks for doing this. Do you need me to merge this for you?

nickdesaulniers commented on libc/src/__support/File/file.cpp:
```diff
@@ -245,6 +266,26 @@ FileIOResult File::read_unlocked(void *data, size_t len) {
+  // Directly copy the data into |data|.
+  cpp::span<uint8_t> dataref(static_cast<uint8_t *>(data) + available_data,
+                             len - available_data);
+  auto result = platform_read(this, dataref.data(), dataref.size());
```
I know the existing code already uses `auto`, but can we just use `FileIOResult`?

jackhong12 commented:
I can submit the code by myself. Let me check how to merge it. Thanks for your reviews!
